src/updater.cpp

```diff
@@ -0,0 +1,187 @@
+// if it's silent updates, do that right away, otherwise
+// setup manual updates to do our custom actions
+void Updater::checkUpdatesOnStart()
+{
+    // only update release builds to prevent auto-update from overwriting your
+    // compiled binary while in development
+    if (mudlet::self()->onDevelopmentVersion()) {
+        //        return;
+    }
+
+#ifdef Q_OS_MACOS
+    setupOnMacOS();
+#endif
+#ifdef Q_OS_LINUX
+    setupOnLinux();
+#endif
+}
+
+#ifdef Q_OS_MACOS
+void Updater::setupOnMacOS()
+{
+    CocoaInitializer initializer;
+    if (mudlet::self()->updateAutomatically()) {
+        updater = new SparkleAutoUpdater("https://feeds.dblsqd.com/MKMMR7HNSP65PquQQbiDIw/release/mac/x86_64/appcast");
+        updater->checkForUpdates();
+    }
+}
+#endif
+
+void Updater::setupOnLinux()
+{
+    feed = new dblsqd::Feed("https://feeds.dblsqd.com/MKMMR7HNSP65PquQQbiDIw", "release");
+
+    QObject::connect(feed, &dblsqd::Feed::ready, [=]() { qWarning() << "Checked for updates:" << feed->getUpdates().size() << "update(s) available"; });
+
+    QObject::connect(feed, &dblsqd::Feed::ready, [=]() {
+        if (!mudlet::self()->updateAutomatically()) {
+            return;
+        }
+
+        auto updates = feed->getUpdates();
+        if (updates.isEmpty()) {
+            return;
+        }
+        feed->downloadRelease(updates.first());
+    });
+
+    QObject::connect(feed, &dblsqd::Feed::downloadFinished, [=]() {
+        // if automatic updates are enabled, and this isn't a manual check, perform the automatic update
+        if (!(mudlet::self()->updateAutomatically() && updateDialog->isHidden())) {
+            return;
+        }
+
+        QFuture<void> future = QtConcurrent::run(this, &Updater::untarOnLinux, feed->getDownloadFile()->fileName());
+
+        // replace current binary with the unzipped one
+        auto watcher = new QFutureWatcher<void>;
+        connect(watcher, &QFutureWatcher<void>::finished, this, &Updater::updateBinaryOnLinux);
+        watcher->setFuture(future);
+    });
+
+    // constructing the UpdateDialog triggers the update check
+    updateDialog = new dblsqd::UpdateDialog(feed, mudlet::self()->updateAutomatically() ? dblsqd::UpdateDialog::Manual : dblsqd::UpdateDialog::OnLastWindowClosed);
+    installOrRestartButton = new QPushButton(tr("Update"));
+    updateDialog->addInstallButton(installOrRestartButton);
+    connect(updateDialog, &dblsqd::UpdateDialog::installButtonClicked, this, &Updater::installOrRestartClicked);
+}
+
+void Updater::untarOnLinux(const QString& fileName) const
+{
+    QProcess tar;
+    tar.setProcessChannelMode(QProcess::MergedChannels);
+    // we can assume tar to be present on a Linux system. If it's not, it'd be rather broken.
+    // tar output folder has to end with a slash
+    tar.start("tar", QStringList() << "-xvf" << fileName << "-C" << QStandardPaths::writableLocation(QStandardPaths::TempLocation) + QStringLiteral("/"));
+    if (!tar.waitForFinished()) {
```

Comment:
Is the default 30 seconds wait (will freeze the main GUI if this is in the main thread) the right amount here?

Reply:
Yes - because this is never run on the main GUI thread ;)

Reply:
I've added an assert to ensure it does not.